File: android/app/src/main/java/io/github/cidy02/kudos/backup/UnsignedDeletionReview.kt
package io.github.cidy02.kudos.backup

import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow

/**
 * D8 notify-on-use + anomaly hold. Populated by [BackupRepository] after a
 * successful merge; [io.github.cidy02.kudos.app.KudosApp] presents the
 * review dialog / digest.
 */
object UnsignedDeletionReview {
    data class Hold(
        val titles: List<String>,
        val workIds: List<String>,
        val sourceLabel: String
    ) {
        val count: Int get() = titles.size
    }

    private val pendingHoldState = MutableStateFlow<Hold?>(null)
    private val pendingDigestState = MutableStateFlow<String?>(null)

    val pendingHold: StateFlow<Hold?> = pendingHoldState.asStateFlow()
    val pendingDigest: StateFlow<String?> = pendingDigestState.asStateFlow()

    fun record(summary: BackupRestoreSummary, source: String) {
        if (summary.unsignedHidesHeld >= BackupMergeService.UNSIGNED_HIDE_HOLD_FLOOR) {
            pendingHoldState.value = Hold(
                titles = summary.unsignedHideTitles,
                workIds = summary.heldUnsignedHideWorkIDs,
                sourceLabel = source
            )
            pendingDigestState.value = null
        } else if (summary.unsignedHidesApplied > 0) {
            val applied = summary.unsignedHidesApplied
            pendingDigestState.value =
                "$applied work${if (applied == 1) "" else "s"} moved to Recently Deleted by $source."
            pendingHoldState.value = null
        }
    }

    fun dismissHold() {
        pendingHoldState.value = null
    }

    fun dismissDigest() {
        pendingDigestState.value = null
    }

    fun resetForTests() {
        pendingHoldState.value = null
        pendingDigestState.value = null
    }
}

File: kudos-ao3-reader/Services/UnsignedDeletionReview.swift
import Foundation
import Observation
import SwiftData

/// D8 notify-on-use + anomaly hold. Populated by `KudosBackupService.restore`
/// after a successful merge; `ContentView` presents the review sheet / digest.
@MainActor
@Observable
final class UnsignedDeletionReview {
    static let shared = UnsignedDeletionReview()

    struct Hold: Equatable, Identifiable {
        let id = UUID()
        var titles: [String]
        var workIDs: [UUID]
        var sourceLabel: String

        var count: Int { titles.count }
    }

    var pendingHold: Hold?
    var pendingDigest: String?

    static func record(_ summary: KudosBackupRestoreSummary, source: String) {
        if summary.unsignedHidesHeld >= KudosBackupService.unsignedHideHoldFloor {
            shared.pendingHold = Hold(
                titles: summary.unsignedHideTitles,
                workIDs: summary.heldUnsignedHideWorkIDs,
                sourceLabel: source
            )
            shared.pendingDigest = nil
        } else if summary.unsignedHidesApplied > 0 {
            let applied = summary.unsignedHidesApplied
            shared.pendingDigest =
                "\(applied) work\(applied == 1 ? "" : "s") moved to Recently Deleted by \(source)."
            shared.pendingHold = nil
        }
    }

    func dismissHold() {
        pendingHold = nil
    }

    func dismissDigest() {
        pendingDigest = nil
    }

    func confirmHold(in context: ModelContext) {
        guard let hold = pendingHold else { return }
        KudosBackupService.applyHeldUnsignedHides(workIDs: hold.workIDs, in: context)
        pendingHold = nil
    }

    /// Tests reset the singleton so suites don't leak hold state.
    func resetForTests() {
        pendingHold = nil
        pendingDigest = nil
    }
}
